### ISSUE TITLE
401 Token Expired for Onedrive when syncing/copying a long time

### ISSUE THREAD

**Issue by landall** [rclone/rclone#3490]
#### What is the problem you are having with rclone?

I am syncing about 10TB( need about 15 days) data from an onedrive account to another. 
When I sync it for the second time (no data copy occurs), the token is expired after about one hour.

#### What is your rclone version (output from `rclone version`)

rclone v1.47.0
- os/arch: linux/amd64
- go version: go1.11.5

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

CentOS Linux release 7.6.1810 x64 

####  Which cloud storage system are you using? (eg Google Drive)

Onedrive for business

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs
rclone sync od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs

**Comment by landall** [comment]
I think the problem is that rclone renew the token between copy actions but not renew it between list actions.

**Comment by Cnly** [comment]
I think your guess is correct according to

https://github.com/rclone/rclone/blob/e2b5ed6c7ad9222eb1f38cfb3a4cc8421e1c34ef/lib/oauthutil/renew.go#L41-L52

But to confirm, could you look into the log and see if there's something like `Token expired but no uploads in progress - doing nothing`? Thanks.

**Comment by landall** [comment]
No, just suddenly 401.

2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/cst/Documents/StarCraft II/Saves/Multiplayer: Making directory
2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/Default/AppData/Roaming/Microsoft/Windows/Recent: Making directory
2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/cst/Documents/Tencent Files/QPlus/Temp/scoped_dir6719: Making directory
2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/cst/Documents/StarCraft II/Accounts/0/5-S2-0-1/Saves/Campaign: Making directory
2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/cst/Pictures/iPod Photo Cache/F25: Making directory
2019-09-02 10:16:46 DEBUG : HP-Envy4-Clone/C/Users/cst/Pictures/iPod Photo Cache/F31: Making directory
2019-09-02 10:16:46 DEBUG : 00000000000/D: Making directory
2019-09-02 10:16:46 DEBUG : One drive root '': failed to copy 7 directories
2019-09-02 10:16:46 DEBUG : One drive root '': copied 326 directories
2019-09-02 10:16:46 ERROR : Attempt 3/3 failed with 8 errors and: 401 Unauthorized:
Transferred:           756 / 756 Bytes, 100%, 0 Bytes/s, ETA 0s
Errors:                 8 (retrying may help)
Checks:            359631 / 359631, 100%
Transferred:            0 / 0, -
Elapsed time:    53m43.9s

**Comment by Cnly** [comment]
It seems that we're missing something here, because each time a request is made rclone should refresh the token if needed, which should be relatively independent from the renew timer. @ncw Any thoughts?

**Comment by ncw** [comment]
> It seems that we're missing something here, because each time a request is made rclone should refresh the token if needed, which should be relatively independent from the renew timer. @ncw Any thoughts?

Yes every request is an opportunity to refresh the token.

We know the token is out of date by the timestamp, so this relies on the user's time and timezone being set correctly.

@landall can you check the time and timezone on your computer is correct?

Also do you get any messages like this in the log?

> Time may be set wrong - time from %q is %v different from this computer

**Comment by landall** [comment]
The log is too long and I haven't save it to file...
let me set up another testing to save it.

I guess the timestamp is ok because the expiring of the token is not triggered if some file copying occurs. 
I suggest you to check the file list process, the full listing of my Onedrive account needs many hours without any token renew is triggered. I guess this is the situation.

**Comment by landall** [comment]
Some information about the source onedrive account:
Quota: 25TB
Used: 9.06TB
Files: 119844
Folders: 5424

**Comment by ncw** [comment]
> I suggest you to check the file list process, the full listing of my Onedrive account needs many hours without any token renew is triggered. I guess this is the situation.

Each one of those file lists should be generating an HTTP transaction which can renew the token.

It only if transactions take longer than 1 hour that the problem occurs.

You can see the http transactions with `-vv --dump headers`

**Comment by landall** [comment]
> > It seems that we're missing something here, because each time a request is made rclone should refresh the token if needed, which should be relatively independent from the renew timer. @ncw Any thoughts?
> 
> Yes every request is an opportunity to refresh the token.
> 
> We know the token is out of date by the timestamp, so this relies on the user's time and timezone being set correctly.
> 
> @landall can you check the time and timezone on your computer is correct?
> 
> Also do you get any messages like this in the log?
> 
> > Time may be set wrong - time from %q is %v different from this computer

rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv | tee rclone401.log
grep wrong rclone401.log
grep timestamp rclone401.log

and get nothing

**Comment by ncw** [comment]
> rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv | tee rclone401.log
> grep wrong rclone401.log
> grep timestamp rclone401.log

You'll need 2>&1 to get the log to go into the tee

    rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv 2>&1 | tee rclone401.log

**Comment by landall** [comment]
> > rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv | tee rclone401.log
> > grep wrong rclone401.log
> > grep timestamp rclone401.log
> 
> You'll need 2>&1 to get the log to go into the tee
> 
> ```
> rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv 2>&1 | tee rclone401.log
> ```

let me try again.

**Comment by landall** [comment]
> > rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv | tee rclone401.log
> > grep wrong rclone401.log
> > grep timestamp rclone401.log
> 
> You'll need 2>&1 to get the log to go into the tee
> 
> ```
> rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --progress --create-empty-src-dirs -vv 2>&1 | tee rclone401.log
> ```

nothing wrong again

**Comment by landall** [comment]
> only if transactions take longer than 1 hour that the problem occurs.
> 
> You can see the http transactions with `-vv --dump headers`

which string should i grep?

**Comment by ncw** [comment]
> > only if transactions take longer than 1 hour that the problem occurs.
> > You can see the http transactions with `-vv --dump headers`
> 
> which string should i grep?

You can't really grep this unfortunately...  The HTTP transactions will be obvious in the logs.  You might want to set `--stats 1m` so you get a regular log (don't use --progress).  Then see if you get an HTTP transaction that takes longer than 60 minutes.

If you want you can post the log or if you prefer email it to me nick@craig-wood.com (please put a link to the issue in the email).

**Comment by landall** [comment]
> > > only if transactions take longer than 1 hour that the problem occurs.
> > > You can see the http transactions with `-vv --dump headers`
> > 
> > 
> > which string should i grep?
> 
> You can't really grep this unfortunately... The HTTP transactions will be obvious in the logs. You might want to set `--stats 1m` so you get a regular log (don't use --progress). Then see if you get an HTTP transaction that takes longer than 60 minutes.
> 
> If you want you can post the log or if you prefer email it to me [nick@craig-wood.com](mailto:nick@craig-wood.com) (please put a link to the issue in the email).

I get a log of 153MB, is it ok?

rclone copy od-admin-lolitaup:/ od-a-soulseal:/ --create-empty-src-dirs -vv  --dump headers --stats 1m  2>&1  | tee rclone401header.log

**Comment by ncw** [comment]
You'd better zip the log. It should compress nicely!

**Comment by landall** [comment]
> You'd better zip the log. It should compress nicely!

Only 8MB now. I have sent the email from my hotmail.

**Comment by landall** [comment]
it is solved in master branch, isn't it?

**Comment by XVilka** [comment]
@landall please retry with the latest master branch, specifically after the https://github.com/rclone/rclone/commit/22937e8982e2cebee1b7d955be4f7469d8d4df52 commit.

**Comment by ncw** [comment]
I'm going to close this now as we haven't seen this problem for a while.
